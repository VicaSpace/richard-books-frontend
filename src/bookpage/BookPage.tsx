import { Button, Container, Heading } from '@chakra-ui/react'
import React, { useState } from 'react'
import { Book } from '../components/bookHorizontalSlider/BookHorizontalSlider'
import DATA from '../homepage/data'
import NavBar from '../navbar/NavBar'
import BookCard from './BookCard'

const BookPage: React.FC<{}> = () => {
  const ratedBooks: Book[] = DATA.books
  const [books, setBooks] = useState(ratedBooks)

  const displayAllRatedBooks = (): void => {
    setBooks(ratedBooks)
  }

  const displayBookmarkedBooks = (): void => {
    const bookmarkedIdsStr: string = localStorage.getItem('bookmarked') ?? '[]'
    const bookmarkedIds: number[] = JSON.parse(bookmarkedIdsStr)
    setBooks(ratedBooks.filter((book) => bookmarkedIds.includes(book.id)))
  }

  const toggleBookmark = (id: number): void => {
    const bookmarkedIdsStr: string = localStorage.getItem('bookmarked') ?? '[]'
    const bookmarkedIds: number[] = JSON.parse(bookmarkedIdsStr)
    if (bookmarkedIds.includes(id)) {
      const newBookmarkedIds = bookmarkedIds.filter((e) => e !== id)
      localStorage.setItem('bookmarked', JSON.stringify(newBookmarkedIds))
      // setBooks(ratedBooks.filter((book) => newBookmarkedIds.includes(book.id)))
    } else {
      bookmarkedIds.push(id)
      localStorage.setItem('bookmarked', JSON.stringify(bookmarkedIds))
      // setBooks(ratedBooks.filter((book) => bookmarkedIds.includes(book.id)))
    }
  }

  return (
    <Container>
      <Heading>My Books</Heading>
      <Button onClick={displayAllRatedBooks}>Rated Books</Button>
      <Button onClick={displayBookmarkedBooks}>Bookmarked</Button>
      <ul>
        {books.map((book) => (
          <li key={book.id}>
            <BookCard book={book} onBookmarkClick={toggleBookmark} />
          </li>
        ))}
      </ul>
      <NavBar />
    </Container>
  )
}

export default BookPage
