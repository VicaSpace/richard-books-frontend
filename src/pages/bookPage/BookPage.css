.bookpage-container {
  padding-top: 2.75rem;
}

.bookpage-container h1 {
  margin-block-start: 0;
  font-size: 1rem;
  line-height: 1.22rem;
  font-weight: 400;
  text-align: center;
  margin-bottom: 1.2rem;
}
