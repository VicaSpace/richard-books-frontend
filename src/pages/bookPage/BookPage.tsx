import React, { useEffect, useState } from 'react'
import { Book } from '../../App'
import BookVerticalList from '../../components/bookVerticalList/BookVerticalList'
import ToggleSelector from '../../components/toggleSelector/ToggleSelector'
import NavBar from '../../utils/sharedComponents/navBar/NavBar'
import './BookPage.css'

const BookPage: React.FC<{ books: Book[] }> = ({ books }) => {
  const [displayBooks, setDisplayBooks] = useState(books)
  const [displayAll, setDisplayAll] = useState(true)

  useEffect(() => setDisplayBooks(books), [books])

  const onRatedBookClick = (): void => {
    setDisplayAll(true)
    setDisplayBooks(books)
  }

  const onBookmarkedClick = (): void => {
    setDisplayAll(false)
    const strBookmarkIds = localStorage.getItem('bookmark') ?? '[]'
    const bookmarkIds: number[] = JSON.parse(strBookmarkIds)
    const bookmarkedBooks = books.filter(book => bookmarkIds.includes(book.id))
    setDisplayBooks(bookmarkedBooks)
  }

  const onBookmarkIconClick = (id: number): void => {
    const strBookmarkIds = localStorage.getItem('bookmark') ?? '[]'
    const bookmarkIds: number[] = JSON.parse(strBookmarkIds)
    if (bookmarkIds.includes(id)) {
      const newBookmarkIds = bookmarkIds.filter(e => e !== id)
      localStorage.setItem('bookmark', JSON.stringify(newBookmarkIds))
    } else {
      bookmarkIds.push(id)
      localStorage.setItem('bookmark', JSON.stringify(bookmarkIds))
    }
  }

  return (
    <div className="bookpage-container">
      <h1>My Books</h1>
      <ToggleSelector
        displayAll={displayAll}
        onRatedBookClick={onRatedBookClick}
        onBookmarkedClick={onBookmarkedClick}
      />
      <BookVerticalList
        displayBooks={displayBooks}
        onBookmarkIconClick={onBookmarkIconClick}
      />
      <NavBar />
    </div>
  )
}

export default BookPage
