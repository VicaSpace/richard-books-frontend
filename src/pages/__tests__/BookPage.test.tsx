import React from 'react'
import { render } from '@testing-library/react'
import { MemoryRouter, Route, Routes } from 'react-router'
import BookPage from '../bookPage/BookPage'

const books = [
  {
    id: 1,
    name: 'Book 1',
    img: 'img src',
    genre: 'Children',
    description: 'Description',
    price: 5,
    Author: {
      id: 1,
      name: 'John',
      img: "John's image"
    }
  },
  {
    id: 2,
    name: 'Book 2',
    img: 'img src',
    genre: 'Children',
    description: 'Description',
    price: 5,
    Author: {
      id: 1,
      name: 'John',
      img: "John's image"
    }
  }
]

describe('test book page', () => {
  it('should render correctly', () => {
    const { container } = render(
      <MemoryRouter initialEntries={['/books']}>
        <Routes>
          <Route path="/books" element={<BookPage books={books} />} />
        </Routes>
      </MemoryRouter>
    )
    expect(container).toMatchSnapshot()
  })
})
