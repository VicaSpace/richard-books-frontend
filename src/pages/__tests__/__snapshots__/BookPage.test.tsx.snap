// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test book page should render correctly 1`] = `
<div>
  <div
    class="bookpage-container"
  >
    <h1>
      My Books
    </h1>
    <div
      class="toggle-selector-container"
    >
      <button
        class="toggle-option-button-selected"
      >
        Rated Books
      </button>
      <button
        class="toggle-option-button-not-selected"
      >
        Bookmarked
      </button>
    </div>
    <div
      class="book-vertical-list-container"
    >
      <div
        class="book-vertical-item-container"
      >
        <div
          class="book-vertical-item-img"
        >
          <img
            alt="Book 1"
            src="img src"
          />
        </div>
        <div
          class="book-vertical-item-middle"
        >
          <div
            class="book-vertical-item-rating"
          >
            <svg
              class="book-vertical-item-star"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 1024 1024"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 0 0 .6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0 0 46.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3z"
              />
            </svg>
             5.0
          </div>
          <div
            class="book-vertical-item-name"
          >
            Book 1
          </div>
          <div
            class="book-vertical-item-author"
          >
            John
          </div>
        </div>
        <div
          class="book-vertical-item-right"
        >
          <div
            class="book-vertical-item-bookmark-container"
          >
            <svg
              class="book-vertical-item-bookmark"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g>
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M5 2h14a1 1 0 0 1 1 1v19.143a.5.5 0 0 1-.766.424L12 18.03l-7.234 4.536A.5.5 0 0 1 4 22.143V3a1 1 0 0 1 1-1z"
                />
              </g>
            </svg>
          </div>
          <div>
            <span
              class="book-vertical-item-emphasize"
            >
              $
              5
            </span>
            .00
          </div>
        </div>
      </div>
      <div
        class="book-vertical-item-container"
      >
        <div
          class="book-vertical-item-img"
        >
          <img
            alt="Book 2"
            src="img src"
          />
        </div>
        <div
          class="book-vertical-item-middle"
        >
          <div
            class="book-vertical-item-rating"
          >
            <svg
              class="book-vertical-item-star"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 1024 1024"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 0 0 .6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0 0 46.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3z"
              />
            </svg>
             5.0
          </div>
          <div
            class="book-vertical-item-name"
          >
            Book 2
          </div>
          <div
            class="book-vertical-item-author"
          >
            John
          </div>
        </div>
        <div
          class="book-vertical-item-right"
        >
          <div
            class="book-vertical-item-bookmark-container"
          >
            <svg
              class="book-vertical-item-bookmark"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g>
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M5 2h14a1 1 0 0 1 1 1v19.143a.5.5 0 0 1-.766.424L12 18.03l-7.234 4.536A.5.5 0 0 1 4 22.143V3a1 1 0 0 1 1-1z"
                />
              </g>
            </svg>
          </div>
          <div>
            <span
              class="book-vertical-item-emphasize"
            >
              $
              5
            </span>
            .00
          </div>
        </div>
      </div>
    </div>
    <div
      class="navbar-container"
    >
      <img
        class="navbar-icon"
        src="home-dimmed.svg"
      />
      <img
        class="navbar-icon"
        src="search-dimmed.png"
      />
      <img
        class="navbar-icon"
        src="book.svg"
      />
    </div>
  </div>
</div>
`;
